--- make.inc	2017-03-15 11:36:45.343057823 +0100
+++ make.inc	2017-03-15 11:39:55.082679220 +0100
@@ -17,9 +17,9 @@
 #  should not compile LAPACK with flags such as -ffpe-trap=overflow.
 #
 FORTRAN  = gfortran 
-OPTS     = -O2 -frecursive
+OPTS     = -O2 -frecursive -fPIC
 DRVOPTS  = $(OPTS)
-NOOPT    = -O0 -frecursive
+NOOPT    = -O0 -frecursive -fPIC
 LOADER   = gfortran
 LOADOPTS =
 #
@@ -52,7 +52,7 @@
 # CC is the C compiler, normally invoked with options CFLAGS.
 #
 CC = gcc
-CFLAGS = -O3
+CFLAGS = -O3 -fPIC
 #
 #  The archiver and the flag(s) to use when building archive (library)
 #  If you system has no ranlib, set RANLIB = echo.
