--- makefile	2018-01-19 10:52:03.224863162 +0100
+++ makefile	2018-01-19 10:55:55.377933626 +0100
@@ -1,7 +1,8 @@
 LIBNAME = lpeg
-LUADIR = ../lua/
+LUADIR = ../luajit/include/luajit-2.1
+LUALIB = ../luajit/lib/libluajit-5.1.a
 
-COPT = -O2
+COPT = -O3
 # COPT = -DLPEG_DEBUG -g
 
 CWARNS = -Wall -Wextra -pedantic \
@@ -29,11 +30,14 @@
 
 # For Linux
 linux:
-	make lpeg.so "DLLFLAGS = -shared -fPIC"
+	make lpeg.a "DLLFLAGS = -shared -fPIC"
 
 # For Mac OS
 macosx:
-	make lpeg.so "DLLFLAGS = -bundle -undefined dynamic_lookup"
+	make lpeg.a "DLLFLAGS = -bundle -undefined dynamic_lookup"
+
+lpeg.a: $(FILES)
+	env $(AR) -r lib$@ $(FILES)
 
 lpeg.so: $(FILES)
 	env $(CC) $(DLLFLAGS) $(FILES) -o lpeg.so
@@ -44,7 +48,7 @@
 	./test.lua
 
 clean:
-	rm -f $(FILES) lpeg.so
+	rm -f $(FILES) lpeg.so liblpeg.a
 
 
 lpcap.o: lpcap.c lpcap.h lptypes.h
--- re.lua	2018-01-19 11:02:39.745760941 +0100
+++ re.lua	2018-01-19 11:26:14.684838882 +0100
@@ -1,5 +1,8 @@
 -- $Id: re.lua,v 1.44 2013/03/26 20:11:40 roberto Exp $
 
+-- Documentation:
+-- http://www.inf.puc-rio.br/~roberto/lpeg/re.html
+
 -- imported functions and modules
 local tonumber, type, print, error = tonumber, type, print, error
 local setmetatable = setmetatable
@@ -256,4 +259,5 @@
 
 if version == "Lua 5.1" then _G.re = re end
 
-return re
+-- return re
+return { regex = re } -- MAD
