diff --git a/src/lib_ffi.c b/src/lib_ffi.c
index 07f52df..fa69391 100644
--- a/src/lib_ffi.c
+++ b/src/lib_ffi.c
@@ -791,6 +791,30 @@ LJLIB_CF(ffi_metatype)
   return 1;
 }
 
+LJLIB_CF(ffi_getmetatable)
+{
+  CTState *cts = ctype_cts(L);
+  GCcdata *cd = ffi_checkcdata(L, 1);
+  CType *ct = ctype_raw(cts, cd->ctypeid);
+  GCtab *t = cts->miscmap;
+  cTValue *mt;
+  TValue tv;
+  if (ctype_isref(ct->info) || ctype_isptr(ct->info)) {
+    ct = ctype_rawchild(cts, ct);
+  }
+  if ((ctype_isstruct(ct->info) || ctype_iscomplex(ct->info) ||
+       ctype_isvector(ct->info) )) {
+    setnumV(&tv, (lua_Number) (int32_t) -ctype_typeid(cts, ct));
+    mt = lj_tab_get(L, t, &tv);
+    if (tvistab(mt)) {
+      settabV(L, L->top-1, tabV(mt));
+      return 1;
+    }
+  }
+  setnilV(L->top-1);
+  return 1;
+}
+
 LJLIB_PUSH(top-7) LJLIB_SET(!)  /* Store reference to finalizer table. */
 
 LJLIB_CF(ffi_gc)	LJLIB_REC(.)
